errbot/backends/slack.py

```diff
@@ -565,6 +565,24 @@ def send_message(self, mess):
                 "to %s: %s" % (to_humanreadable, mess.body)
             )
 
+    def send_stream_request(self, identifier, fsource, name=None, size=None, stream_type=None):
+        """Starts a file transfer. For Slack, the size and stream_type are unsupported"""
+        name = name if name else 'file'
+        stream = Stream(identifier, fsource, name, size, stream_type)
+        log.debug('Initiating upload of {0}'.format(name))
+        resp = self.api_call('files.upload', data={
```

Comment:
It looks totally synchronous, maybe enqueue a WorkRequest on the internal threadpool ?
Something along that (untested code):

```python
def upload(stream):
     stream.accept()
     resp = self.api_call('files.upload', data={'channels': stream.identifier.channelid, 'filename': stream.name, 'file': stream)
      log.debug('Upload response: {0}'.format(resp))
      if "ok" in resp and resp["ok"]:
        stream.success()
      else:
        stream.error()

and in send_stream_request:
[...]
self.thread_pool.putRequest(WorkRequest(self.upload, [stream, ]))
return stream

```